File: python/aether/custom_metals.py
from mcresources import ResourceManager

rm = ResourceManager("modpack")

metals = [
    {"metal": "aluminium", "fluid": "kubejs:aluminium"},
    {"metal": "aluminium_bronze", "fluid": "createbigcannons:molten_bronze"},
    {"metal": "harderned_steel", "fluid": "createbigcannons:molten_steel"},
    {"metal": "cast_iron_alloy", "fluid": "createbigcannons:molten_cast_iron"},
    {"metal": "nethersteel", "fluid": "createbigcannons:molten_nethersteel"}
]

def mold_filling(metal, fluid):
    rm.recipe("custom_metals/mold_filling/mold/"+ metal,"create:filling",{
    "ingredients":[
        {"item": "tfc:ceramic/ingot_mold",
        "nbt": ""},
        {"fluid": fluid,
         "amount": 100}
    ],
    "results": [
       {"item": "tfc:ceramic/ingot_mold",
        "nbt": {"tank":{"Amount":100,"FluidName": fluid}}},
    ]
    })
    rm.recipe("custom_metals/mold_filling/fire_mold/"+ metal,"create:filling",{
    "ingredients":[
        {"item": "tfc:ceramic/fire_ingot_mold",
        "nbt": ""},
        {"fluid": fluid,
         "amount": 100}
    ],
    "results": [
       {"item": "tfc:ceramic/fire_ingot_mold",
        "nbt": {"tank":{"Amount":100,"FluidName": fluid}}},
    ]
})

def welding(path, metal, input_type, output_type):
    rm.recipe("custom_metals/welding/" + path,"tfc:welding", {
    "first_input": {
        "item": "kubejs:metal/" + input_type + "/" + metal
    },
    "second_input": {
        "item": "kubejs:metal/" + input_type + "/" + metal
    },
    "tier": 0,
    "result": {
        "item": "kubejs:metal/" + output_type + "/" + metal
    }
})

def rolling(path, metal):
    rm.recipe("custom_metals/rolling/" + path,"createaddition:rolling",{
	"input": {
      	"item": "kubejs:metal/ingot/" + metal
	},
	"result": {
		"item": "kubejs:metal/rod/" + metal,
		"count": 2
	}
})

def fluid_compacting(path, metal, output_type, fluid, fluid_amount):
    rm.recipe("custom_metals/compacting/" + path,"create:compacting",{
    "ingredients": [
      {
        "amount": fluid_amount,
        "fluid": fluid
      }
    ],
    "results": [
      {
        "item": "kubejs:metal/"+ output_type + "/" + metal
      }
    ]
})
    
def item_compacting(path, metal, input_type, output_type):
    rm.recipe("custom_metals/compacting/" + path,"create:compacting",{
    "ingredients": [
        {
            "item": "kubejs:metal/" + input_type + "/" + metal
        },
        {
            "item": "tfc:powder/flux"
        },
        {
            "item": "kubejs:metal/" + input_type + "/" + metal
        }
    ],
    "results": [
        {
          "item": "kubejs:metal/" + output_type + "/" + metal
        }
    ],
    "heatRequirement": "heated"
})

def cutting(path, metal, input_type, output_type):
    rm.recipe("custom_metals/cutting/" + path, "create:cutting",{
    "ingredients": [
        {
        "item": "kubejs:metal/" + input_type + "/" + metal
        }
    ],
    "processingTime": 50,
    "results": [
        {
        "item": "kubejs:metal/" + output_type + "/" + metal,
        "count": 2
        }
    ]
})
    
def pressing(path, metal):
    rm.recipe("custom_metals/pressing/" + path, "create:pressing", {
    "ingredients": [
    {
    "item": "kubejs:metal/double_ingot/" + metal
    }
    ],
    "results": [
    {
    "item": "kubejs:metal/sheet/" + metal
    }
    ]
})

def sheet_anvil(path, metal):
    rm.recipe("custom_metals/anvil/sheet/" + path, "tfc:anvil", {
    "input": {
        "item": "kubejs:metal/double_ingot/" + metal
    },
    "result": {
        "item": "kubejs:metal/sheet/" + metal
    },
    "tier": 2,
    "rules": [
        "hit_last",
        "hit_second_last",
        "hit_third_last"
    ]
})

def rod_anvil(path, metal):
    rm.recipe("custom_metals/anvil/rod/" + path, "tfc:anvil", {
    "input": {
        "item": "kubejs:metal/ingot/" + metal
    },
    "result": {
        "item": "kubejs:metal/rod/" + metal,
        "count": 2
    },
    "tier": 2,
    "rules": [
        "bend_last",
        "draw_second_last",
        "draw_third_last"
    ]
})

def generate():
    for m in metals:
        metal = m.get("metal")
        fluid = m.get("fluid")
        cutting("double_ingot/" + metal, metal, "double_ingot", "ingot")

        cutting("double_sheet/" + metal, metal, "double_sheet", "double_ingot")
        welding("double_ingot/" + metal, metal, "ingot", "double_ingot")
        item_compacting("double_ingot/" + metal, metal, "ingot", "double_ingot")

        fluid_compacting("sheet/" + metal, metal, "sheet", fluid, 200)
        pressing("sheet/" + metal, metal)
        sheet_anvil("sheet/" + metal, metal)

        item_compacting("double_sheet/" + metal, metal, "sheet", "double_sheet")
        welding("double_sheet/" + metal, metal, "sheet", "double_sheet")

        rod_anvil("rod/" + metal, metal)
        rolling("rod/" + metal, metal)



generate()

File: kubejs/server_scripts/datagen/custom_metals.js
//priority: 1
ServerEvents.recipes(event =>{

    global.KUBEJS_METALS.forEach(metal => {

        event.recipes.tfc.welding(Item.of("kubejs:metal/double_ingot/" + metal), Item.of("kubejs:metal/ingot/" + metal), Item.of("kubejs:metal/ingot/" + metal))
            .id("modpack:welding/" + metal + "_double_ingot")
        
        
        event.recipes.tfc.welding(Item.of("kubejs:metal/double_sheet/" + metal), Item.of("kubejs:metal/sheet/" + metal), Item.of("kubejs:metal/sheet/" + metal))
            .id("modpack:welding/" + metal + "_double_sheet")
        
        event.recipes.tfc.anvil("kubejs:metal/sheet/" + metal, "kubejs:metal/double_ingot/" + metal, [
            "hit_last",
            "hit_second_last",
            "hit_third_last"
        ]).tier(2).id("modpack:anvil/sheet/" + metal)

        event.recipes.tfc.anvil(Item.of("kubejs:metal/rod/" + metal, 2), "kubejs:metal/ingot/" + metal, [
            "bend_last",
            "draw_second_last",
            "draw_third_last"
        ]).tier(2).id("modpack:anvil/rod/" + metal)

        event.recipes.tfc.casting("kubejs:metal/ingot/" + metal, "tfc:ceramic/ingot_mold", Fluid.of(global.METAL_FLUIDS[metal], 100), 0.1)
            .id("modpack:casting/ingot_mold/" + metal)
        
        event.recipes.tfc.casting("kubejs:metal/ingot/" + metal, "tfc:ceramic/fire_ingot_mold", Fluid.of(global.METAL_FLUIDS[metal], 100), 0.01)
            .id("modpack:casting/fire_ingot_mold/" + metal)
        
        event.recipes.tfc.heating("kubejs:metal/ingot/" + metal, global.MELTING_POINTS[metal]).resultFluid(Fluid.of(global.METAL_FLUIDS[metal], 100)).id("modpack:heating/ingot/" + metal)
        event.recipes.tfc.heating("kubejs:metal/double_ingot/" + metal, global.MELTING_POINTS[metal]).resultFluid(Fluid.of(global.METAL_FLUIDS[metal], 200)).id("modpack:heating/double_ingot/" + metal)
        event.recipes.tfc.heating("kubejs:metal/sheet/" + metal, global.MELTING_POINTS[metal]).resultFluid(Fluid.of(global.METAL_FLUIDS[metal], 200)).id("modpack:heating/sheet/" + metal)
        event.recipes.tfc.heating("kubejs:metal/double_sheet/" + metal, global.MELTING_POINTS[metal]).resultFluid(Fluid.of(global.METAL_FLUIDS[metal], 400)).id("modpack:heating/double_sheet/" + metal)
        event.recipes.tfc.heating("kubejs:metal/rod/" + metal, global.MELTING_POINTS[metal]).resultFluid(Fluid.of(global.METAL_FLUIDS[metal], 50)).id("modpack:heating/rod/" + metal)
    })
})

TFCEvents.data(event =>{
    
    global.KUBEJS_METALS.forEach(metal => {
        event.metal(global.METAL_FLUIDS[metal], global.MELTING_POINTS[metal], 0.02143, 
            "kubejs:metal/ingot/" + metal, "kubejs:metal/double_ingot/" + metal, "kubejs:metal/sheet/" + metal, 2, "modpack:" + metal
        )
    })
})
